# 문제 1 -----------------------------------------------------------------------------------
감기약 위약을 투여받은 환자의 부작용은, 항암약 위약을 투여받은 환자의 부작용과 차이가 있는가?

##### 문제 1-1 
항암약을 투여받은 환자 중 '이상없음'의 비율을 0과 1사이로 구하시오

In [1]:
# 항암약 부작용
import pandas as pd
df = pd.DataFrame({ 
    "항암약" : [4,4,3,4,1,4,1,4,1,4,4,2,1,4,2,3,2,4,4,4]
})

In [5]:
print(df.head(10))

   항암약
0    4
1    4
2    3
3    4
4    1
5    4
6    1
7    4
8    1
9    4


# [!] value_counts에 normalize=True 옵션을 설정하면, 각 값의 상대적 비율로 반환된다.

In [2]:
#답
df_counts = df.value_counts(normalize=True)
print(df_counts)
df_len = len(df)
print(df_len)

항암약
4      0.55
1      0.20
2      0.15
3      0.10
dtype: float64
20


In [23]:
# 답
result = (df_counts.values[0]/df_len)
print(result)

0.55


##### 문제 1-2
감기약의 예상 부작용 비율과 항암약의 부작용 관찰값이 통계적으로 유의미하게 차이가 있는지 확인하려 한다.
카이제곱 검정을 사용해 검정 통계량을 구하시오.

In [ ]:
#부작용 비율이라고 하면. 따로 넣어줘야 하나?

In [ ]:
# 감기약의 부작용 비율 

In [25]:
a = [0.1, 0.05, 0.15, 0.7]
b = [4/20, 3/20, 2/20, 11/20]

In [27]:
#두 집단의 분포? 확인 -> 적합도?
from scipy import stats
result = stats.chi2_contingency(a,b)

In [35]:
from scipy.stats import chisquare

expected_value = a # 여기에 각 20을 곱해줘야 한다.
expected_value = [x * 20 for x in expected_value]
observed_value = df.value_counts().sort_index().to_list()

result = chisquare(f_obs = observed_value, f_exp =expected_value)
print(result)


Power_divergenceResult(statistic=6.976190476190476, pvalue=0.07266054733847573)


##### 문제 1-3
위의 p-value를 구하시오. => 0.07266054733847573

# 문제 2 -----------------------------------------------------------------------------------
다중 선형 회귀 모델을 구축하고, 각 소문제에 답하시오.

In [37]:
import pandas as pd
df = pd.read_csv("data6-3-2.csv")
print(df.head(3))

    solar  wind     o3  temperature
0   89.14  6.28  33.52         23.0
1  109.97  1.04  27.01         20.7
2  102.83  6.42  41.00         20.5


##### 문제 2-1 
주어진 데이터를 기반으로 다중 선형 회귀 모델을 구축하고, 독립변수 o3의 회귀 계수를 구하시오

In [39]:
from statsmodels.formula.api import ols
model = ols('temperature~solar+wind+o3', data=df).fit()

In [40]:
model.params['o3']

0.07493854378136569

##### 문제 2-2 
문제 2-1에서 적합한 모델에서 solar와 o3이 고정된 상태에서 wind의 세기가 증가함에 따라 temperature가 감소한다.
이 회귀 모델에서 wind의 회귀 계수에 대한 p-value를 구하시오.

In [42]:
model.pvalues['wind']

0.7797177202071734

##### 문제 2-3
문제 2-1에서 적합한 모델에서 solar:100, wind:5, o3:30일 때, 예측값을 구하시오.

# [!] 이번에 확실히 알고 가기

In [43]:
example = {'solar' :[100], 'wind' :[5], 'o3' :[30]}
pred = model.predict(example)

In [45]:
pred[0]

21.561630469479688